doc/manual.asciidoc

```diff
@@ -187,10 +187,57 @@ Ninja defaults to running commands in parallel anyway, so typically
 you don't need to pass `-j`.)
 
 
+GNU Jobserver support
+~~~~~~~~~~~~~~~~~~~~~
+
+Since version 1.13., Ninja builds support the
+https://https://www.gnu.org/software/make/manual/html_node/Job-Slots.html[GNU Make jobserver]
+protocol (on Posix systems). If supports both client and
+server modes.
+
+Client mode is useful when Ninja is invoked as part of a larger
+build system controlled by a top-level GNU Make instance, as it
+allows better coordination between concurrent build tasks.
+
+Server mode is useful when Ninja is the top-level build tool that
+invokes sub-builds recursively in a similar setup.
+
+To enable server mode, use `--jobserver` or `--jobserver-mode=MODE`
+on the command line, or set `NINJA_JOBSERVER=MODE` in your
+environment, where `MODE` can be one of the following values:
+
+`0`: Do not enable the feature (the default)
+`1`: Enable the feature, using the best mode for the current system.
+`pipe`: Enable the feature, implemented with an anonymous pipe (Posix only).
+`fifo`: Enable the feature, implemented with a FIFO file path (Posix only).
+`sem`: Enable the feature, implemented with a Win32 semaphore (Windows only).
+
+Note that `--jobserver` is equivalent to `--jobserver-mode=1`.
+
+Otherwise, the client feature is automatically enabled for builds
+(not tools) under the following conditions:
+
+- Dry-run (i.e. `-n` or `--dry-run`) is not enabled.
+
+- `-j1` (no parallelism) is not used on the command line.
```

Comment:
But as a *client* wouldn't couldn't I still be limited by jobserver? I.e. waiting on a token to become available. But I guess that isn't supported yet as nothing will wake up ninjas main loop in that case?

Reply:
The possible behaviors are the following *when* a jobserver pool is in place in `MAKEFLAGS`:

A) `-j1` disables parallel tasks and ignores the pool
B) `-j2` and above (also `-j0`) ignore the command-line job count, and use the pool to control parallelism.
C) Without a command-line `-j` parameter, use the pool.

D) `-j1` is ignored, Ninja uses the pool to control parallelism.
E) `-j2` and above (also `-j0`) ignore the pool and use the command-line job count instead.

The current PR implements A) and B) because `-j1` is the only way to tell Ninja that we do not want parallelism, and the only reason to do that would be for debugging a build or because the system is very constrained (e.g. not enough RAM). Hence, ignoring it if a jobserver pool is in place seems unhelpful.

The reason B) exists (explicit job counts are ignored when the pool is in place) is because many build scripts will invoke Ninja with an explicit count, oblivious to the fact that a jobserver is in place. Doing this allows everything to work transparently without modifying tons of configuration files or scripts in complex multi-build systems.

In theory, you can setup a pool with a depth of 1, but since each client receives an implicit job slot when it starts, whether `-j1` ignores the pool or not results in exactly the same behavior whether A) or D) is implemented.

Reply:
A) sounds pretty reasonable; but I think B) should generate a warning.

Reply:
FWIW, there is already an `Info()` message printed in this case "jobserver detected ...", unless `--quiet` is specified.

Reply:
I think I favor C (if I understand it correctly that it means: with `-j` always ignore the pool).

1. It's simple. Less code, less documentation.
2. It's intuitive: When I specify a job count, I would want ninja to respect that and not override the behavior based on an environment variable.
3. We don't change behavior with an update. ninja 1.12 uses 42 jobs if you pass `-j42`, ninja 1.13 should too.
4. We *do* change the behavior without `-j` - but that currently already means "auto detect" and I think it's expectable for that auto detect mechanism to improve in an update.
5. People might currently use `-j<some low number>` as a workaround to calm ninja because it doesn't respect the jobserver. We shouldn't encourage them to keep a workaround in, so encourage the removal.